crates/subspace-service/src/lib.rs

```diff
@@ -261,104 +261,110 @@ where
 
+                    let parent_slot = parent_pre_digest.slot();
+                    if slot <= *parent_slot {
                         return false;
                     }
-                };
-
-                let parent_slot = parent_pre_digest.slot();
-                if slot <= *parent_slot {
-                    return false;
-                }
 
-                let pot_parameters = match client.runtime_api().pot_parameters(parent_hash) {
-                    Ok(pot_parameters) => pot_parameters,
-                    Err(error) => {
-                        debug!(
-                            %error,
-                            %parent_hash,
-                            parent_number = %parent_header.number(),
-                            "Failed to retieve proof of time parameters during proof of time \
-                            verification"
-                        );
+                    let pot_parameters = match client.runtime_api().pot_parameters(parent_hash) {
+                        Ok(pot_parameters) => pot_parameters,
+                        Err(error) => {
+                            debug!(
+                                %error,
+                                %parent_hash,
+                                parent_number = %parent_header.number(),
+                                "Failed to retrieve proof of time parameters during proof of time \
+                                verification"
+                            );
 
-                        return false;
-                    }
-                };
+                            return false;
+                        }
+                    };
 
-                let pot_input = if parent_header.number().is_zero() {
-                    PotNextSlotInput {
-                        slot: parent_slot + Slot::from(1),
-                        slot_iterations: pot_parameters.slot_iterations(),
-                        seed: pot_verifier.genesis_seed(),
+                    let pot_input = if parent_header.number().is_zero() {
+                        PotNextSlotInput {
+                            slot: parent_slot + Slot::from(1),
+                            slot_iterations: pot_parameters.slot_iterations(),
+                            seed: pot_verifier.genesis_seed(),
+                        }
+                    } else {
+                        let pot_info = parent_pre_digest.pot_info();
+
+                        PotNextSlotInput::derive(
+                            pot_parameters.slot_iterations(),
+                            parent_slot,
+                            pot_info.proof_of_time(),
+                            &pot_parameters.next_parameters_change(),
+                        )
+                    };
+
+                    // Ensure proof of time and future proof of time included in upcoming block are
+                    // valid
+
+                    if quick_verification {
+                        tokio::task::block_in_place(|| {
```

Comment:
This could be much simplified IMO by using a https://docs.rs/tokio/latest/tokio/runtime/struct.Builder.html#method.new_current_thread (which could be created ahead of time and reused serially)

Reply:
That is a builder to create a brand new runtime. But we don't want a new runtime here, we want to do a blocking operation in place (API that Substrate exposes is blocking and there is nothing we can do about it, this particular one realistically will not become async any time soon), while avoiding blocking other async tasks and `tokio::task::block_in_place` is designed exactly for this use case.

Reply:
I meant ahead of time .. at L262 above, create the runtime and capture in the closure. This nested `block_in_place()/block_on()` are hard to follow

Reply:
I do not understand why we need a new runtime there either, especially because we have access to tokio multi-threaded runtime there as well. All `extensions_for` does is creates a few extensions that are potentially called from block execution via WASM host functions later.

Reply:
What I don't understand is: this is a sync fn/closure, what guarantees this gets called from an async context? If it were async, sure there won't be any doubts..

Reply:
It is not necessarily called from async context, but it is called from a thread with tokio runtime. For example `tokio::task::spawn_blocking` creates a perfectly synchronous thread, but it has tokio context, so you can use usualy tokio stuff there, including `Handle::current().block_on()` and such.

Code here is executed from within one of `tokio::task::spawn_blocking()` calls that Substrate initiated, just through multiple levels of abstractions.